indico/modules/events/timetable/util.py

```diff
@@ -0,0 +1,158 @@
+from collections import defaultdict
+
+from pytz import utc
+from sqlalchemy.orm import joinedload, subqueryload
+
+from indico.core.db import db
+from indico.modules.events.contributions.models.contributions import Contribution
+from indico.modules.events.models.events import Event
+from indico.modules.events.models.persons import EventPersonLink
+from indico.modules.events.sessions.models.blocks import SessionBlock
+from indico.modules.events.sessions.models.sessions import Session
+from indico.modules.events.timetable.models.breaks import Break
+from indico.modules.events.timetable.models.entries import TimetableEntry
+from indico.util.date_time import iterdays
+
+
+def is_visible_from(event, categ):
+    """Check whether ``event`` is visible from ``categ``
+    """
+    visibility = event.as_legacy.getFullVisibility()
+    for cat_id in event.category_chain:
+        if visibility <= 0:
+            return False
+        if str(cat_id) == categ.id:
+            return True
+        visibility -= 1
+    return True
+
+
+def _query_events(categ_ids, day_start, day_end):
+    event = db.aliased(Event)
+    dates_overlap = lambda t: (t.start_dt >= day_start) & (t.start_dt <= day_end)
+    return (db.session.query(Event.id, TimetableEntry.start_dt)
+            .filter(
+                Event.category_chain.overlap(categ_ids),
+                ~Event.is_deleted,
+                ((Event.timetable_entries.any(dates_overlap(TimetableEntry))) |
```

Comment:
I think complex nested criteria that don't fit on one line end up more readable when using `db.and_` and `db.or_` instead of the binary operators.

```python
def foo():
    return (db.session.query(Event.id, TimetableEntry.start_dt)
            .filter(
                Event.category_chain.overlap(categ_ids),
                ~Event.is_deleted,
                db.or_(
                    Event.timetable_entries.any(dates_overlap(TimetableEntry)),
                    Event.query.exists().where(Event.happens_between(day_start, day_end) & (Event.id == event.id))
                )
            )
            .group_by(Event.id, TimetableEntry.start_dt)
            .order_by(Event.id, TimetableEntry.start_dt)
            .join(TimetableEntry, (TimetableEntry.event_id == Event.id) & (dates_overlap(TimetableEntry)),
                  isouter=True))
```